Segfault in tests/call_stack_args.b on x86_64 Linux

After #110 I'm seeing a segfault on x86_64 in one of the stack argument tests:

``` console
$ make build/tests/call_stack_args
build/b -run -o build/tests/call_stack_args tests/call_stack_args.b ./std/test.b
Generated build/tests/call_stack_args.asm
[INFO] CMD: fasm build/tests/call_stack_args.asm build/tests/call_stack_args.o
flat assembler  version 1.73.27  (16384 kilobytes memory)
3 passes, 1440 bytes.
[INFO] CMD: cc -no-pie -o build/tests/call_stack_args build/tests/call_stack_args.o
[INFO] CMD: build/tests/call_stack_args
[ERROR] command process was terminated by signal 11
make: *** [Makefile:169: build/tests/call_stack_args] Error 1
```

Not sure what platform info is relevant here, but here's `neofetch` (which I learned about on the stream btw):

``` console
$ neofetch
            .-/+oossssoo+/-.               mikko@sl 
        `:+ssssssssssssssssss+:`           -------- 
      -+ssssssssssssssssssyyssss+-         OS: Ubuntu 22.04.1 LTS on Windows 10 x86_64 
    .ossssssssssssssssssdMMMNysssso.       Kernel: 6.6.87.1-microsoft-standard-WSL2 
   /ssssssssssshdmmNNmmyNMMMMhssssss/      Uptime: 17 hours, 48 mins 
  +ssssssssshmydMMMMMMMNddddyssssssss+     Packages: 583 (dpkg) 
 /sssssssshNMMMyhhyyyyhmNMMMNhssssssss/    Shell: bash 5.1.16 
.ssssssssdMMMNhsssssssssshNMMMdssssssss.   Terminal: Windows Terminal 
+sssshhhyNMMNyssssssssssssyNMMMysssssss+   CPU: Intel i5-1035G7 (8) @ 1.497GHz 
ossyNMMMNyMMhsssssssssssssshmmmhssssssso   GPU: a0d3:00:00.0 Microsoft Corporation Device 008e 
ossyNMMMNyMMhsssssssssssssshmmmhssssssso   Memory: 711MiB / 3710MiB 
+sssshhhyNMMNyssssssssssssyNMMMysssssss+
.ssssssssdMMMNhsssssssssshNMMMdssssssss.                           
 /sssssssshNMMMyhhyyyyhdNMMMNhssssssss/                            
  +sssssssssdmydMMMMMMMMddddyssssssss+
   /ssssssssssshdmNNNNmyNMMMMhssssss/
    .ossssssssssssssssssdMMMNysssso.
      -+sssssssssssssssssyyyssss+-
        `:+ssssssssssssssssss+:`
            .-/+oossssoo+/-.
```

I'll try to debug it a further myself too once I have more time, but thought I'd report for now.

--- a/src/codegen/fasm_x86_64.rs
+++ b/src/codegen/fasm_x86_64.rs
@@ -207,61 +207,46 @@ pub unsafe fn generate_function(name: *const c_char, params_count: usize, auto_v
                 }
             }
             Op::Funcall{result, fun, args} => {
+                let mut i = 0;
+                while i < cmp::min(args.count, registers.len()) {
+                    let reg = (*registers)[i];
+                    load_arg_to_reg(*args.items.add(i), reg, output);
+                    i += 1;
+                }
+
+                // args on the stack are push right to left
+                // so we need to iterate them in reverse
+                let mut push_count = 0;
+                for j in (i..args.count).rev() {
+                    load_arg_to_reg(*args.items.add(j), c!("rax"), output);
+                    sb_appendf(output, c!("    push rax\n"));
+                    push_count += 1;
+                }
+                if push_count % 2 != 0 {
+                    sb_appendf(output, c!("    push rax\n")); // Align stack
+                    push_count += 1;
+                }
+
                 match os {
                     Os::Linux => {
-                        let mut i = 0;
-                        while i < cmp::min(args.count, registers.len()) {
-                            let reg = (*registers)[i];
-                            load_arg_to_reg(*args.items.add(i), reg, output);
-                            i += 1;
-                        }
-
-                        // args on the stack are push right to left
-                        // so we need to iterate them in reverse
-                        let mut push_count = 0;
-                        for j in (i..args.count).rev() {
-                            load_arg_to_reg(*args.items.add(j), c!("rax"), output);
-                            sb_appendf(output, c!("    push rax\n"));
-                            push_count += 1;
-                        }
-                        if push_count % 2 != 0 {
-                            sb_appendf(output, c!("    push rax\n")); // Align stack
-                            push_count += 1;
-                        }
-
                         sb_appendf(output, c!("    mov al, 0\n")); // x86_64 Linux ABI passes the amount of
                                                                    // floating point args via al. Since B
                                                                    // does not distinguish regular and
                                                                    // variadic functions we set al to 0 just
                                                                    // in case.
                         call_arg(fun, output);
-
-                        if push_count > 0 {
-                            sb_appendf(output, c!("    add rsp, %zu\n"), push_count*8);
-                        }
                     }
                     Os::Windows => {
-                        let mut i = 0;
-                        while i < cmp::min(args.count, registers.len()) {
-                            let reg = (*registers)[i];
-                            load_arg_to_reg(*args.items.add(i), reg, output);
-                            i += 1;
-                        }
-
-                        // args on the stack are push right to left
-                        // so we need to iterate them in reverse
-                        for j in (i..args.count).rev() {
-                            load_arg_to_reg(*args.items.add(j), c!("rax"), output);
-                            sb_appendf(output, c!("    push rax\n"));
-                        }
-
                         // allocate 32 bytes for "shadow space"
                         // it must be allocated at the top of the stack after all arguments are pushed
                         // so we can't allocate it at function prologue
                         sb_appendf(output, c!("    sub rsp, 32\n"));
                         call_arg(fun, output);
-                        sb_appendf(output, c!("    add rsp, %zu\n"), (args.count-i)*8+32); // deallocate stack args & "shadow space"
+                        sb_appendf(output, c!("    add rsp, 32\n")); // deallocate "shadow space"
                     }
+                }
+                if push_count > 0 {
+                    sb_appendf(output, c!("    add rsp, %zu\n"), push_count*8); // deallocate stack args
                 }
                 sb_appendf(output, c!("    mov [rbp-%zu], rax\n"), result*8);
             },

--- a/src/codegen/fasm_x86_64.rs
+++ b/src/codegen/fasm_x86_64.rs
@@ -218,7 +218,7 @@ pub unsafe fn generate_function(name: *const c_char, params_count: usize, auto_v
                 if stack_args_count > 0 {
                     sb_appendf(output, c!("    sub rsp, %zu\n"), stack_args_size);
                     for i in 0..stack_args_count {
-                        load_arg_to_reg(*args.items.add(i), c!("rax"), output);
+                        load_arg_to_reg(*args.items.add(reg_args_count + i), c!("rax"), output);
                         sb_appendf(output, c!("    mov [rsp+%zu], rax\n"), i*8);
                     }
                 }

--- a/src/codegen/fasm_x86_64.rs
+++ b/src/codegen/fasm_x86_64.rs
@@ -218,9 +218,15 @@ pub unsafe fn generate_function(name: *const c_char, params_count: usize, auto_v
 
                         // args on the stack are push right to left
                         // so we need to iterate them in reverse
+                        let mut push_count = 0;
                         for j in (i..args.count).rev() {
                             load_arg_to_reg(*args.items.add(j), c!("rax"), output);
                             sb_appendf(output, c!("    push rax\n"));
+                            push_count += 1;
+                        }
+                        if push_count % 2 != 0 {
+                            sb_appendf(output, c!("    push rax\n")); // Align stack
+                            push_count += 1;
                         }
 
                         sb_appendf(output, c!("    mov al, 0\n")); // x86_64 Linux ABI passes the amount of
@@ -229,7 +235,10 @@ pub unsafe fn generate_function(name: *const c_char, params_count: usize, auto_v
                                                                    // variadic functions we set al to 0 just
                                                                    // in case.
                         call_arg(fun, output);
-                        sb_appendf(output, c!("    add rsp, %zu\n"), (args.count-i)*8); // deallocate stack args
+
+                        if push_count > 0 {
+                            sb_appendf(output, c!("    add rsp, %zu\n"), push_count*8);
+                        }
                     }
                     Os::Windows => {
                         let mut i = 0;

--- a/src/codegen/fasm_x86_64.rs
+++ b/src/codegen/fasm_x86_64.rs
@@ -207,18 +207,17 @@ pub unsafe fn generate_function(name: *const c_char, params_count: usize, auto_v
                 }
             }
             Op::Funcall{result, fun, args} => {
-                let mut i = 0;
-                while i < cmp::min(args.count, registers.len()) {
+                let reg_args_count = cmp::min(args.count, registers.len());
+                for i in 0..reg_args_count {
                     let reg = (*registers)[i];
                     load_arg_to_reg(*args.items.add(i), reg, output);
-                    i += 1;
                 }
 
                 // args on the stack are push right to left
                 // so we need to iterate them in reverse
                 let mut push_count = 0;
-                for j in (i..args.count).rev() {
-                    load_arg_to_reg(*args.items.add(j), c!("rax"), output);
+                for i in (reg_args_count..args.count).rev() {
+                    load_arg_to_reg(*args.items.add(i), c!("rax"), output);
                     sb_appendf(output, c!("    push rax\n"));
                     push_count += 1;
                 }

--- a/src/codegen/fasm_x86_64.rs
+++ b/src/codegen/fasm_x86_64.rs
@@ -213,19 +213,15 @@ pub unsafe fn generate_function(name: *const c_char, params_count: usize, auto_v
                     load_arg_to_reg(*args.items.add(i), reg, output);
                 }
 
-                // args on the stack are push right to left
-                // so we need to iterate them in reverse
-                let mut push_count = 0;
-                if args.count > reg_args_count && args.count % 2 != 0 {
-                    sb_appendf(output, c!("    push rax\n")); // Align stack
-                    push_count += 1;
+                let stack_args_count = args.count - reg_args_count;
+                let stack_args_size = align_bytes(stack_args_count*8, 16);
+                if stack_args_count > 0 {
+                    sb_appendf(output, c!("    sub rsp, %zu\n"), stack_args_size);
+                    for i in 0..stack_args_count {
+                        load_arg_to_reg(*args.items.add(i), c!("rax"), output);
+                        sb_appendf(output, c!("    mov [rsp+%zu], rax\n"), i*8);
+                    }
                 }
-                for i in (reg_args_count..args.count).rev() {
-                    load_arg_to_reg(*args.items.add(i), c!("rax"), output);
-                    sb_appendf(output, c!("    push rax\n"));
-                    push_count += 1;
-                }
-                assert!(push_count % 2 == 0, "stack is not aligned");
 
                 match os {
                     Os::Linux => {
@@ -245,8 +241,8 @@ pub unsafe fn generate_function(name: *const c_char, params_count: usize, auto_v
                         sb_appendf(output, c!("    add rsp, 32\n")); // deallocate "shadow space"
                     }
                 }
-                if push_count > 0 {
-                    sb_appendf(output, c!("    add rsp, %zu\n"), push_count*8); // deallocate stack args
+                if stack_args_count > 0 {
+                    sb_appendf(output, c!("    add rsp, %zu\n"), stack_args_size);
                 }
                 sb_appendf(output, c!("    mov [rbp-%zu], rax\n"), result*8);
             },

--- a/src/codegen/fasm_x86_64.rs
+++ b/src/codegen/fasm_x86_64.rs
@@ -216,15 +216,16 @@ pub unsafe fn generate_function(name: *const c_char, params_count: usize, auto_v
                 // args on the stack are push right to left
                 // so we need to iterate them in reverse
                 let mut push_count = 0;
+                if args.count > reg_args_count && args.count % 2 != 0 {
+                    sb_appendf(output, c!("    push rax\n")); // Align stack
+                    push_count += 1;
+                }
                 for i in (reg_args_count..args.count).rev() {
                     load_arg_to_reg(*args.items.add(i), c!("rax"), output);
                     sb_appendf(output, c!("    push rax\n"));
                     push_count += 1;
                 }
-                if push_count % 2 != 0 {
-                    sb_appendf(output, c!("    push rax\n")); // Align stack
-                    push_count += 1;
-                }
+                assert!(push_count % 2 == 0, "stack is not aligned");
 
                 match os {
                     Os::Linux => {

--- a/tests/call_stack_args.b
+++ b/tests/call_stack_args.b
@@ -3,6 +3,8 @@
 
 f(a0,a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11) {
     extrn printf;
+    printf("%d %d %d %d %d %d %d %d %d %d\n", a0, a1, a2, a3, a4, a5, a6, a7, a11, a10);
+    printf("%d %d %d %d %d %d %d %d %d %d %d\n", a0, a1, a2, a3, a4, a5, a6, a7, a11, a10, a9);
     printf("%d %d %d %d %d %d %d %d %d %d %d %d\n", a0, a1, a2, a3, a4, a5, a6, a7, a11, a10, a9, a8);
 }
 

--- a/tests/call_stack_args.b
+++ b/tests/call_stack_args.b
@@ -1,5 +1,4 @@
 // TODO: incorrect output on uxn
-// TODO: does not compile on gas-aarch64-linux
 
 f(a0,a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11) {
     extrn printf;

